pkg/connections/stomp/requestor.go

```diff
@@ -0,0 +1,119 @@
+/*
+Copyright (c) 2018 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+  http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package client contains the types and functions used to communicate with other services using
+// queues and topics.
+package stomp
+
+import (
+	"github.com/container-mgmt/messaging-library/pkg/client"
+	"github.com/golang/glog"
+	"github.com/segmentio/ksuid"
+)
+
+// Requestor is an implementation of Requestor interface
+// The stomp requestor is a specification of the connection interface
+type Requestor struct {
+	conn          *Connection
+	destination   string
+	responseQueue string
+}
+
+// NewRequestor creates a new requestor API to submit requests
+func (c *Connection) NewRequestor(spec client.RequestorSpec) (r client.Requestor) {
+	stompRequestor := &Requestor{
+		conn:          c,
+		destination:   spec.Destination,
+		responseQueue: spec.ResponseQueue,
+	}
+
+	r = stompRequestor
+	return
+}
+
+// Send sends a request to a specific destination
+// request is the message request
+// callback is the handler that will be called when the response is received
+// requestID is returned upon successful send
+func (r *Requestor) Send(request client.Message, callback client.ResponseCallback) (requestID string, err error) {
+
+	// This will bw called when the response arrives.
+	// Call the ResponseCallback and unsubscribe
+	subscriptionCallback := func(m client.Message, destination string) (err error) {
+
+		// Check message received with no error
+		if m.Err != nil {
+			err = m.Err
+			return
+		}
+
+		// Validate message is a response
+		if kind, ok := m.Data["kind"]; !ok || kind.(string) != "Response" {
+			// ignore message
+			glog.Warningf(
+				"Message of non 'Response' kind received from destination %s. Ignoring",
+				destination)
+			return
+		}
+
+		// Parse requestId
+		id, ok := m.Data["requestID"]
+		if !ok {
+			// ignore message
+			glog.Warningf(
+				"Response missing 'requestID' field recieved from destination %s. Ignoring",
+				destination)
+			return
+		}
+
+		// Validate requestID
+		if id.(string) != requestID {
+			// ignore message
+			glog.Warningf(
+				"Expected response to request %s. Received request id %s. Ignoring",
+				requestID,
+				id.(string))
+			return
+		}
+
+		// unsubscribe from respondTo
+		err = r.conn.Unsubscribe(r.responseQueue)
+
+		// call the callback
+		callback(m, requestID)
+		return
+	}
+
+	// Subscribe to respondTo with the callback
+	err = r.conn.Subscribe(r.responseQueue, subscriptionCallback)
```

Comment:
I find it confusing the fact that the request/reply mechanism depends on the publish/subscribe mechanism. We may introduce publis/subscribe specific behaviour in the `Subscribe` and `Publish` methods that may not be good for request/reply. Can we use instead directly the methods of the STOMP library?

Reply:
Sure we can. I was trying to make share usage of common code.
Even if we don't call it publish/subscirbe - the implementation will need to use that functionality

Reply:
opened https://github.com/container-mgmt/messaging-library/issues/64 to track this